connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java

```diff
@@ -102,6 +101,62 @@ public static boolean isJar(Path path) {
         return locations;
     }
 
+    public static List<Path> pluginUrls(Path topPath) throws IOException {
+        Set<Path> archives = new HashSet<>();
+        LinkedList<DirectoryEntry> dfs = new LinkedList<>();
+        Set<Path> visited = new HashSet<>();
+
+        if (isArchive(topPath)) {
+            return Collections.singletonList(topPath);
+        }
+
+        DirectoryStream<Path> topListing = Files.newDirectoryStream(
+                topPath,
+                PLUGIN_FILTER
+        );
+        dfs.push(new DirectoryEntry(topListing));
+        visited.add(topPath);
+        try {
+            while (!dfs.isEmpty()) {
+                Iterator<Path> neighbors = dfs.peek().iterator;
+                if (neighbors.hasNext()) {
```

Comment:
That's a good point. Nesting confused me too, that when I contemplated `continue`, I saw the else in the end that pops the item, and thought it would skip it.